src/libmongoc/src/mongoc/mongoc-topology.c

```diff
@@ -1682,3 +1659,165 @@ _mongoc_topology_bypass_cooldown (mongoc_topology_t *topology)
    BSON_ASSERT (topology->single_threaded);
    topology->scanner->bypass_cooldown = true;
 }
+
+static void
+_find_topology_version (const bson_t *reply, bson_t *topology_version)
+{
+   bson_iter_t iter;
+   const uint8_t *bytes;
+   uint32_t len;
+
+   if (!bson_iter_init_find (&iter, reply, "topologyVersion") ||
+       !BSON_ITER_HOLDS_DOCUMENT (&iter)) {
+      bson_init (topology_version);
+      return;
+   }
+   bson_iter_document (&iter, &len, &bytes);
+   bson_init_static (topology_version, bytes, len);
+}
+
+
+/* "Clears" the connection pool by incrementing the generation.
+ * 
+ * Pooled clients with open connections will discover the invalidation
+ * the next time they fetch a stream to the server.
+ * 
+ * Caller must lock topology->mutex. */
+void
+_mongoc_topology_clear_connection_pool (mongoc_topology_t *topology,
+                                        uint32_t server_id)
+{
+   mongoc_server_description_t *sd;
+   bson_error_t error;
+
+   sd = mongoc_topology_description_server_by_id (
+      &topology->description, server_id, &error);
+   if (!sd) {
+      /* Server removed, ignore and ignore error. */
+      return;
+   }
+   TRACE ("clearing pool for server: %s", sd->host.host_and_port);
+   sd->generation++;
+}
+
+
+/* Handle an error from an app connection.
```

Comment:
Arguably, this logic should be in mongoc-cluster.c, which is responsible for creating and managing connections for application use. However, the new SDAM tests mock application errors. I found it simpler to consolidate the logic of the [Application Errors|https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#application-errors] section here, so this can be used by the tests and by mongoc-cluster.c.

Consequently, the spec tests only test the logic of updating the server description and generation on application errors. Shane has integration tests in the works (https://github.com/ShaneHarvey/specifications/pull/1). And libmongoc has some existing test coverage of handling "not master" errors with tests using the mock server.